src/atlas/interpolation/method/sphericalvector/SphericalVector.cc

```diff
@@ -0,0 +1,296 @@
+
+void SphericalVector::do_execute(const FieldSet& sourceFieldSet,
+                                 FieldSet& targetFieldSet,
+                                 Metadata& metadata) const {
+  ATLAS_TRACE("atlas::interpolation::method::SphericalVector::do_execute()");
+
+  const auto nFields = sourceFieldSet.size();
+  ATLAS_ASSERT(nFields == targetFieldSet.size());
+
+  for (auto i = 0; i < sourceFieldSet.size(); ++i) {
+    do_execute(sourceFieldSet[i], targetFieldSet[i], metadata);
+  }
+}
+
+void SphericalVector::do_execute(const Field& sourceField, Field& targetField,
+                                 Metadata&) const {
+  ATLAS_TRACE("atlas::interpolation::method::SphericalVector::do_execute()");
+
+  const auto fieldType = sourceField.metadata().getString("type", "");
+  if (fieldType != "vector") {
+
+    auto metadata = Metadata();
+    Method::do_execute(sourceField, targetField, metadata);
+
+    return;
+  }
+
+  if (target_.size() == 0) {
+    return;
+  }
+
+  ATLAS_ASSERT_MSG(sourceField.variables() == 2 || sourceField.variables() == 3,
+                   "Vector field can only have 2 or 3 components.");
+
+  Method::check_compatibility(sourceField, targetField, matrix());
+
+  haloExchange(sourceField);
+
+  if (sourceField.datatype().kind() == array::DataType::KIND_REAL64) {
+    interpolate_vector_field<double>(sourceField, targetField);
+  } else if (sourceField.datatype().kind() == array::DataType::KIND_REAL32) {
+    interpolate_vector_field<float>(sourceField, targetField);
+  } else {
+    ATLAS_NOTIMPLEMENTED;
+  }
+
+  targetField.set_dirty();
+}
+
+template <typename Value>
+void SphericalVector::interpolate_vector_field(const Field& sourceField,
+                                               Field& targetField) const {
+  if (sourceField.rank() == 2) {
+    interpolate_vector_field<Value, 2>(sourceField, targetField);
+  } else if (sourceField.rank() == 3) {
+    interpolate_vector_field<Value, 3>(sourceField, targetField);
+  } else {
+    ATLAS_NOTIMPLEMENTED;
+  }
+}
+
+template <typename Value, int Rank>
+void SphericalVector::interpolate_vector_field(const Field& sourceField,
+                                               Field& targetField) const {
+
+  const auto sourceView = array::make_view<Value, Rank>(sourceField);
+  auto targetView = array::make_view<Value, Rank>(targetField);
+  targetView.assign(0.);
+
+  const auto horizontalComponent = [](const auto& sourceVars, auto& targetVars,
+                                      const auto& complexWeight) {
+    const auto sourceVector = Complex(sourceVars(0), sourceVars(1));
+    const auto targetVector = complexWeight * sourceVector;
+    targetVars(0) += targetVector.real();
+    targetVars(1) += targetVector.imag();
+  };
+
+  if (sourceField.variables() == 2) {
+    matrixMultiply(sourceView, targetView, horizontalComponent,
+                   *complexWeights_);
+    return;
+  } else if (sourceField.variables() == 3) {
```

Comment:
"else if" should be "if" instead (because of the "return").

Reply:
Clearly I've lazily attempted to refactor at some point. Done.